lib/utils/verify-signatures.js

```diff
@@ -0,0 +1,460 @@
+    const cachedKey = `${registry}-/npm/v1/keys`
+
+    try {
+      const entry = await cacache.get(cachePath, cachedKey)
+      const cache = jsonParse(entry.data)
+      if (cache.expires && Date.now() > Date.parse(cache.expires)) {
+        await cacache.rm.entry(cachePath, cachedKey)
+        throw new Error('Cache expired')
+      } else {
+        return cache.keys
+      }
+    } catch {
+      const keys = await fetch.json('/-/npm/v1/keys', {
+        ...this.npm.flatOptions,
+        registry,
+      }).then(({ keys }) => keys.map((key) => {
+        key.pemKey = `-----BEGIN PUBLIC KEY-----\n${key.key}\n-----END PUBLIC KEY-----`
+        return key
+      })).catch(err => {
+        if (err.code === 'E404') {
+          return null
+        } else {
+          throw err
+        }
+      })
+      const inOneWeekMs = 1000 * 60 * 60 * 24 * 7
+      const cache = {
+        expires: new Date(Date.now() + inOneWeekMs).toISOString(),
+        keys,
+      }
+      await cacache.put(cachePath, cachedKey, JSON.stringify(cache))
+      return keys
+    }
+  }
+
+  async getVerifiedInfo (edge) {
+    let alias = false
+    try {
+      alias = npa(edge.spec).subSpec
+    } catch (err) {
+    }
+    const spec = npa(alias ? alias.name : edge.name)
+    const node = edge.to || edge
+    const { path, location } = node
+    const { version } = node.package || {}
+    if (!version) {
+      return
+    }
+    const type = edge.optional ? 'optionalDependencies'
+      : edge.peer ? 'peerDependencies'
+      : edge.dev ? 'devDependencies'
+      : 'dependencies'
+
+    for (const omitType of this.npm.config.get('omit')) {
+      if (node[omitType]) {
+        return
+      }
+    }
+
+    // deps different from prod not currently
+    // on disk are not included in the output
+    if (edge.error === 'MISSING' && type !== 'dependencies') {
+      return
+    }
+
+    try {
+      const packument = await this.getPackument(spec)
+      // if it's not a range, version, or tag, skip it
+      try {
+        if (!npa(`${edge.name}@${edge.spec}`).registry) {
+          return null
+        }
+      } catch (err) {
+        return null
+      }
+
+      const name = alias ? edge.spec.replace('npm', edge.name) : edge.name
+      const { homepage } = packument
+
+      const registry = fetch.pickRegistry(spec, this.npm.flatOptions)
+
+      const versionPackument = pickManifest(packument, version, this.npm.flatOptions)
+      const versionCreated = packument.time && packument.time[version]
+      const dist = versionPackument.dist || {}
+      const { integrity } = dist
+      const message = `${name}@${version}:${integrity}`
+      const signatures = dist.signatures || []
+      const keys = (await this.getKeys({ registry })) || []
+      const validKeys = keys.filter((publicKey) => {
+        if (!publicKey.expires) {
+          return true
+        }
+        return Date.parse(versionCreated) < Date.parse(publicKey.expires)
```

Comment:
I'm having trouble wrapping my head around this check.

We're already checking if `expires` is in the past.  What is this checking? If this version was created after the publicKey expires?  This would only be possible if the version was published in the future.

Reply:
It could be, if the local system's clock and the registry's clock aren't the same.

Reply:
> If this version was created after the publicKey expires?

Yeah this might be an unlikely edge case but was thinking of the case where we need to backdate an expiry in the past and re-write some signatures. In this scenario you might end up with two keys, one expired/compromised and another legit one and we only want to check the signature that had a valid key at the the time the version was published.